### DISCUSSION TITLE
Understanding the results of multi-dimension motif discovery using MSTUMP

### DISCUSSION THREAD

**Discussion by davejaivik** [TDAmeritrade/stumpy#369] [Q&A]
Based on [Matrix Profile VI](https://www.cs.ucr.edu/~eamonn/Motif_Discovery_ICDM.pdf) Paper and [Stumpy Documentation (Tutorial-Mutlidimensional Motif Discovery)](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html) of the same, I'm trying to identify Multi-dimensional Motifs from  [PRONTO Dataset](https://zenodo.org/record/1341583).

After a bit of preprocessing of the data, and calculating multidimensional matrix profile, I plot the Elbow Curve as suggested in the tutorial. Based on the curve k=8 dimensions out of 31 are relevant. 
![Elbow_31](https://user-images.githubusercontent.com/40367099/111955010-f1f37780-8b0e-11eb-8b36-c7b7c4277101.png)

When I plot the motifs after selecting the varibels using MP subspace, and I get the following result. Now, even though there are aligned motifs, but not all of them are aligned.
![image](https://user-images.githubusercontent.com/40367099/111960425-c0ca7580-8b15-11eb-9c9d-065cb5712671.png)

Even when I try with subset of variables I get misaligned motifs.

**Comment by seanlaw** [comment]
@davejaivik Thank you for the question. I'm not familiar with the PRONTO dataset but can you please provide a minimal-reproducible code example to reproduce all of the work that you are showing above? Otherwise, it's hard for us to speculate as to what is happening.  Additionally, it isn't clear why there are multiple dashed lines and red highlights. Would you mind explaining?

**Reply by davejaivik** [reply]
Hi, Thanks for the reply. I have uploaded [code](https://github.com/davejaivik/PRONTO-Multidimensional-Motifs) with brief explanation of PRONTO and showing the results over it. Red highlighted sequences represent Motifs and black dashed line points to which point in k-dimensional Matrix Profile(Final Graph in the Plot) does the motif correspond to.

**Reply by seanlaw** [reply]
@davejaivik I'm finding it rather hard to follow what it is that you are trying to plot so I shall try my best. From the elbow plot, if you choose `k=7` (i.e., `8` dimensions) then, following the original tutorial, you would compute the subspace with:

```
k = 7
S = stumpy.subspace(df, m, motifs_idx[k][0], ind[k][motifs_idx[k][0]], k)
print(f"For k = {k}, the {k+1}-dimensional subspace includes subsequences from variables/columns: {S}")
``` 

which prints out:

```
For k = 7, the 8-dimensional subspace includes subsequences from variables/columns: [12 17 20 22 23 25 29  7]
```

Thus, it means that for these `8` dimensions you only look at the pairs of subsequences located/starting at two specific locations (i.e., at `motifs_idx[k][0]` and at `ind[k][motifs_idx[k][0]]`; Note that these two locations are dependent on `mps[k]`):

```
f,ax = plt.subplots(k+1, 1, figsize=(20,20),sharex=True)
motif_idx = motifs_idx[k][0]
nearest_neighbor_idx = ind[k][motifs_idx[k][0]]
for i, s in enumerate(S):
    tmp = df[df.columns[s]]
    ax[i].plot(tmp)
    ax[i].set_title("Variable:{}".format(s))
    ax[i].plot(tmp[motif_idx : motif_idx + m], c='orange',linewidth=4)
    ax[i].plot(tmp[nearest_neighbor_idx : nearest_neighbor_idx + m], c='r',linewidth=4)
    ax[i].axvline(x=motif_idx,linestyle='dashed',c='k',linewidth=2)
    ax[i].axvline(x=nearest_neighbor_idx,linestyle='dashed',c='k',linewidth=2)
```

![temp](https://user-images.githubusercontent.com/7473521/112243569-30c72180-8c24-11eb-861a-c6c0cf5690c3.png)

This looks right to me as when you print the the smallest multi-dimensional matrix profile value for `k=7`:

```
print(np.min(mps[7]))
```

You get:

```
2.1582503986385368e-05
```

This near-zero value means that the average distance computed from all `8` pairs of subsequences (taken from dimensions `[12 17 20 22 23 25 29  7]`) is small. Or, in other words, each pair of subsequences are very similar to each other and, when averaged, is still very small. This is consistent with the plot that I generated above. Therefore, I think that the MSTUMP results are correct and accurate.

Additionally, in case there is any confusion, the motif found for, say, `k=0` and `k=5` may be completely unrelated (i.e., they need not be aligned and most likely aren't) as they may be choosing a different set of dimensions from which to compute their averaged matrix profile distance.

**Reply by davejaivik** [reply]
Hi, @seanlaw thank you for the reply. I got the aligned motifs by selecting the starting points of subsequences from mps[k]. But then wouldn't it be possible to get the aligned motifs for any value of k? 

I also tried to verify the results by finding motifs from individual/uni-variate Matrix Profile for all the variables identified by MP Subspace and not all the motifs were aligned or relatively close (I expected them to be similar if not exact to the motifs identified by mstump). I have posted the notebook (PRONTO Multidimensional Motif Discovery - 2) with the results [here](https://github.com/davejaivik/PRONTO-Multidimensional-Motifs).

Fig.1: Aligned Motifs by taking motif indices from mps[k] (k=7 from elbow curve)
![image](https://user-images.githubusercontent.com/40367099/113599825-cae38c80-965c-11eb-9a8f-8419f2ee841e.png)


Fig.2: Individual Motifs for each variable and corresponding Matrix Profile ( Variable are selected based on MP Subspace )
![image](https://user-images.githubusercontent.com/40367099/113599959-f8303a80-965c-11eb-87ed-284c6baf2468.png)
